### ISSUE TITLE
Serious performance issues crawling a huge network disk

### ISSUE THREAD

**Issue by radomirml** [Norconex/collector-filesystem#59]
We're running a network disk crawling process for a few months already.  We're using Norconex library to crawl the disk, filter by metadata and to commit text to Elasticsearch. After a couple of iterations we got rid of the internal Apache Tika based extractor and integrated dtSearch for  text extraction (faster and much lower RAM requirements).

Although everything was smooth in the begining, we're facing serious performance issues. The crawling is so slow that with 6 crawling threads we get only about 1 or 2 files processed every hour! (We still didn't complete crawling of the whole drive.)

I filtered logs to show processing time of a single file - please check the attached screenshot. 

![2021-01-29_19-26 crawler perf](https://user-images.githubusercontent.com/89235/106313717-ec915380-6268-11eb-9e6a-70f5a08cc5d5.png)

You can see additional logging messages that I added to track down the performance issue. The "processed file" time shows time spent in the FileImporterPipeline:

```
		long start = System.currentTimeMillis();
        new FileImporterPipeline(
                getCrawlerConfig().isKeepDownloads(), getCrawlerConfig().isSkipDocumentFetch()).execute(fileContext);
        ImporterResponse res = fileContext.getImporterResponse();
        LOG.info("*** Processed " + crawlData.getReference() + " in " + (System.currentTimeMillis() - start) + "ms");
```

It appears that the file spent 3 hours in the importer's pipeline; and that doesn't even include text extraction and submission to ES. 

More details about the system:

- 6 GB allocated for the Java process
- 8 cores on the system, 50% CPU usage non-stop
- crawler store 180 GB (860K files, 85K folders) of which most goes to the ES committer queue. (Why data is kept after sending it to ES?)

Do you any idea what's going on here? Thanks for any feedback you may have!

**Comment by essiembre** [comment]
Hello @radomirml.  

I have to ask before digging deeper: can you reproduce the issue using the Apache Tika based extractor? We need to rule out the dtSearch integration.  

Data should definitely not be kept once sent to ES. There is something fishy for sure.

A few questions:
* What type of filesystem is this?  Local disk, network, else?  
* Does it do that on huge files only?  What is your average file size?
* Does it always hang on the same files, or is it totally random?
* Does it always hang at around the same number of docs crawled?
* Do you have any log errors?

Typically, the default crawl store starts to suffer performance problems when it reaches the million (not sure I witnessed it getting that slow though).  If that is the case here, my best advice (short of creating/using a distributed crawl store) is to split your job into multiple collectors.  You can reuse config fragments to avoid duplication, or automate your config generation.   The idea is to have different start paths (and different committer queue dir) in each configuration and/or different reference filters.

Each can also be installed on a different machine if need be (for increased performance).

**Comment by radomirml** [comment]
Hello @essiembre ,

Thanks for getting back to me. To answer your questions:

- It's a a network drive, crawled using UNC path.
- It's slow with smaller files as well. Hard to say what's an average file size. We're trying to index about 12TB of content with about 8M of documents. I limited indexed file size to 100MB so we shouldn't be handling anything bigger.
- We already split indexing to 6 systems, each crawling a different subset of folders. 
- The system in my screenshot is at 1.2M files processed. (There was another one though that processed 2.6M and completed.)
- I cannot say if it hangs on the same files. After running it for months, with several restarts, we still cannot fully crawl on all systems. It never completely stops, just processing gets slower and slower.
- The problem is reproducible with the Apache Tika a well. I added dtSearch implementation because it was faster and more memory efficient, but we're still hitting the issue at some point.
- I even disabled file content reading code in Norconex, leaving it to the dtSearch extractor - just to make sure that it's not a problem with file read performance. As the log shows, dtSearch code completed in 15s (which is still a lot for a single file) and extracted 18MB of text. How that takes 3 additional hours to be processed, I don't understand. I suspect on store.
- The sample in the screenshot is just an except from the log but all files at that point were taking so long to process. 

Thanks.

**Comment by essiembre** [comment]
From the information you share, it really seems to point to the default crawlstore limitations (which is an embedded key/value store).   I see a couple of options for you:

1. Have more collectors/crawlers and make sure they each handle fewer documents.
2. Use the `MongoCrawlDataStoreFactory` instead. MongoDB can be distributed as needed to support a higher load.
3. Create your own crawl store by implementing `ICrawlDataStoreFactory`.

**Comment by radomirml** [comment]
Thank you @essiembre

**Comment by radomirml** [comment]
For anyone who might run into a similar issue... I've finally found what was killing the performance. As a part of the crawler configuration, we've enabled the TitleGeneratorTagger. However, the summarize() method of this tagger was eating CPU more than anything else, having very bad performance depending on content size and type. For example:
- 20MB of lorem ipsum text would take more 97 seconds to process
- 20MB of text extracted for a spreadsheet would need 45 minutes to process
In both cases totally unnecessary and unexpected. After disabling this tagger, things started to run much better.
